sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataBufferingOutboundObserver.java

```diff
@@ -43,48 +57,83 @@
+      "beam_fn_api_data_buffer_size_limit=";
+  public static final String BEAM_FN_API_DATA_BUFFER_TIME_LIMIT =
+      "beam_fn_api_data_buffer_time_limit=";
   @VisibleForTesting static final int DEFAULT_BUFFER_LIMIT_BYTES = 1_000_000;
+  private static final long DEFAULT_BUFFER_LIMIT_TIME_MS = -1L;
   private static final Logger LOG =
       LoggerFactory.getLogger(BeamFnDataBufferingOutboundObserver.class);
 
   public static <T> BeamFnDataBufferingOutboundObserver<T> forLocation(
+      PipelineOptions options,
       LogicalEndpoint endpoint,
       Coder<T> coder,
       StreamObserver<BeamFnApi.Elements> outboundObserver) {
-    return forLocationWithBufferLimit(
-        DEFAULT_BUFFER_LIMIT_BYTES, endpoint, coder, outboundObserver);
+    return new BeamFnDataBufferingOutboundObserver<>(
+        getSizeLimit(options), getTimeLimit(options), endpoint, coder, outboundObserver);
   }
 
-  public static <T> BeamFnDataBufferingOutboundObserver<T> forLocationWithBufferLimit(
-      int bufferLimit,
-      LogicalEndpoint endpoint,
-      Coder<T> coder,
-      StreamObserver<BeamFnApi.Elements> outboundObserver) {
-    return new BeamFnDataBufferingOutboundObserver<>(
-        bufferLimit, endpoint, coder, outboundObserver);
+  private static int getSizeLimit(PipelineOptions options) {
+    List<String> experiments = options.as(ExperimentalOptions.class).getExperiments();
+    for (String experiment : experiments == null ? Collections.<String>emptyList() : experiments) {
+      if (experiment.startsWith(BEAM_FN_API_DATA_BUFFER_SIZE_LIMIT)) {
+        return Integer.parseInt(experiment.substring(BEAM_FN_API_DATA_BUFFER_SIZE_LIMIT.length()));
+      }
+      if (experiment.startsWith(BEAM_FN_API_DATA_BUFFER_LIMIT)) {
+        return Integer.parseInt(experiment.substring(BEAM_FN_API_DATA_BUFFER_LIMIT.length()));
+      }
+    }
+    return DEFAULT_BUFFER_LIMIT_BYTES;
+  }
+
+  private static long getTimeLimit(PipelineOptions options) {
+    List<String> experiments = options.as(ExperimentalOptions.class).getExperiments();
+    for (String experiment : experiments == null ? Collections.<String>emptyList() : experiments) {
+      if (experiment.startsWith(BEAM_FN_API_DATA_BUFFER_TIME_LIMIT)) {
+        return Long.parseLong(experiment.substring(BEAM_FN_API_DATA_BUFFER_TIME_LIMIT.length()));
+      }
+    }
+    return DEFAULT_BUFFER_LIMIT_TIME_MS;
   }
 
   private long byteCounter;
   private long counter;
   private boolean closed;
-  private final int bufferLimit;
+  private final int sizeLimit;
   private final Coder<T> coder;
   private final LogicalEndpoint outputLocation;
   private final StreamObserver<BeamFnApi.Elements> outboundObserver;
   private final ByteString.Output bufferedElements;
+  @VisibleForTesting final ScheduledFuture<?> flushFuture;
+  private final AtomicReference<IOException> flushException = new AtomicReference<>(null);
+  private final Object lock = new Object();
 
   private BeamFnDataBufferingOutboundObserver(
```

Comment:
I think we will want to have the original  implementation of this class so that we don't need to take on the synchronization overhead for flush.